drivers/audio/wm8904.c

```diff
@@ -0,0 +1,728 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <errno.h>
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/audio/codec.h>
+#include <zephyr/devicetree/clocks.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(wolfson_wm8904);
+
+#include "wm8904.h"
+
+#define DT_DRV_COMPAT wolfson_wm8904
+
+struct wm8904_driver_config {
+	struct i2c_dt_spec i2c;
+	int clock_source;
+	const struct device *mclk_dev;
+	clock_control_subsys_t mclk_name;
+	const struct pinctrl_dev_config *pincfg;
+};
+
+#define DEV_CFG(dev) ((const struct wm8904_driver_config *const)dev->config)
+
+static void wm8904_write_reg(const struct device *dev, uint8_t reg, uint16_t val);
+static void wm8904_read_reg(const struct device *dev, uint8_t reg, uint16_t *val);
+static void wm8904_update_reg(const struct device *dev, uint8_t reg, uint16_t mask, uint16_t val);
+static void wm8904_soft_reset(const struct device *dev);
+
+static void wm8904_configure_output(const struct device *dev);
+
+static void wm8904_configure_input(const struct device *dev);
+
+static int wm8904_initialize(const struct device *dev)
+{
+	LOG_DBG("Initializing codec.");
+
+	const struct wm8904_driver_config *const dev_cfg = DEV_CFG(dev);
+
+	int error = pinctrl_apply_state(dev_cfg->pincfg, PINCTRL_STATE_DEFAULT);
+
+	if (error) {
+		LOG_ERR("pinctrl_apply_state fail: %d", error);
+		return error;
+	}
+
+	error = i2c_recover_bus(dev_cfg->i2c.bus);
```

Comment:
Why does this driver always need to recover the I2C bus?

Reply:
@anangl Looks like a requirement when trying to use RT595/RT685's I3C peripheral as an I2C master.